libmisc/strlcpy.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2022, Alejandro Colomar <alx.manpages@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the copyright holders or contributors may not be used to
+ *    endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stddef.h>
+
+#ident "$Id$"
+
+#include "prototypes.h"
+
+
+/*
+ * ARGUMENTS
+ *	dst	Pointer to the first byte of the destination buffer.
+ *
+ *	src	Pointer to the first byte of the source string.
+ *
+ *	n	Size of the 'dst' buffer.
+ *
+ * DESCRIPTION
+ *	This function reads the whole 'str' string, until a NUL
+ *	character is found, and copies it (or the part of it that fits)
+ *	into 'dst', guaranteeing that the resulting string is
+ *	NUL-terminated.
+ *
+ * RETURN VALUE
+ *	This function returns strlen(src).
+ *
+ * CAVEATS
+ *	This function is difficult to use for concatenating strings,
+ *	since the programmer needs to recalculate the 'dst' size every
+ *	time.  strlcat(3bsd) solves that issue, but it's unnecessarily
+ *	slow, since it reads the whole 'dst' string, to find the NUL
+ *	character.  stpecpy() is both easy to use and fast.
+ *
+ *	This function requires that the input string is NUL-terminated.
+ *	If this cannot be guaranteed, other functions should be used,
+ *	such as strscpy().
+ */
+
+size_t
+strlcpy(char *restrict dst, const char *restrict src, size_t n)
+{
+	char    *d = dst;
+	size_t  i;
+
+	for (i = 0; src[i] != '\0'; i++) {
+		if (i < n)
+			*d++ = src[i];
+	}
+	if (d - dst < n)
+		*d = '\0';
```

Comment:
I don't think this is right.  If n = 2, and strlen(src) > 2, then *d here will (if i'm thinking right) be pointing at *(d[2]).

Reply:
Yeah, you're right; I was off by one.  I guess it should be
```c
if (n)
        dst[n - 1] = '\0';
```

Reply:
BTW, I would wait to merge the agetpass() thing, and then we can just pick strlcpy(3bsd) from libbsd.